File: src/main/java/arcana/common/recipes/ArcanaRecipes.java
package arcana.common.recipes;

import arcana.Arcana;
import net.minecraft.item.crafting.IRecipe;
import net.minecraft.item.crafting.IRecipeSerializer;
import net.minecraft.item.crafting.SpecialRecipeSerializer;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;

import static arcana.utils.Util.arcLoc;

@SuppressWarnings("unchecked")
public class ArcanaRecipes {

    public static class Serializers{
        public static IRecipeSerializer<WandsRecipe> CRAFTING_WANDS = withName(new SpecialRecipeSerializer<>(WandsRecipe::new), "wands");


        private static <T extends IRecipe<?>> IRecipeSerializer<T> withName(IRecipeSerializer<T> serializer, String regName){
            return (IRecipeSerializer<T>) serializer.setRegistryName(arcLoc(regName));
        }
    }
}

File: src/main/kotlin/arcana/common/recipes/ArcanaRecipes.kt
package arcana.common.recipes

import arcana.utils.Util
import net.minecraft.item.crafting.IRecipe
import net.minecraft.item.crafting.IRecipeSerializer
import net.minecraft.item.crafting.SpecialRecipeSerializer

class ArcanaRecipes {
    object Serializers {
        @JvmField val CRAFTING_WANDS = withName(SpecialRecipeSerializer { id -> WandsRecipe(id) }, "wands")
        private fun <R : IRecipe<*>> withName(serializer: IRecipeSerializer<R>, regName: String): IRecipeSerializer<R> {
            return serializer.setRegistryName(Util.arcLoc(regName)) as IRecipeSerializer<R>
        }
    }
}